application-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml

```diff
@@ -390,6 +400,29 @@ define('xwiki-ckeditor', [
     return oldInline.call(this, element, ckeditor.tools.extend(getConfig(element), config, true));
   };
 
+  // Define a way to switch on or off the temporary attachment upload mechanism.
+  ckeditor.switchTemporaryAttachmentUpload = function (isEnabled, instance, config) {
```

Comment:
I don't understand the need for this. If someone wants to disable the temporary attachment upload for a specific CKEditor instance they can do it:
* either by passing an instance configuration object when creating the editor, if they control the editor creation
```
ckeditor.replace(textarea, {
  'xwiki-upload': {
    'isTemporaryAttachmentSupported': false
  }
});
```
* or by listening to CKEditor events as indicated on https://extensions.xwiki.org/xwiki/bin/view/Extension/CKEditor+Integration#HConfiguretheEditor and changing the config after it is loaded.

Reply:
I wanted to provide the capability of disabling the feature even through the advanced WYSIWYG configuration in XWiki, as a failsafe mechanism to disable it globally if needed. AFAIU the options you present here, wouldn't work there. 
Now maybe it was a bit of overengineering to provide this: we can always document that users can edit EditSheet for example to disable it, or any other sheet including the editor. Or programmatically like you described.

Reply:
If we want to allow users to disable the usage of temporary attachments in CKEditor globally, for all editor instances, then we need to make this configuration global. Currently this configuration is implemented as instance level configuration: the value is read from the HTML element that is enhanced by the CKEditor (``data-temporary-attachment-supported``), so two elements can use different values, thus is cannot be overwritten globally. In order to make it a global config we need to:

* stop reading it from the editor element
* instead, read it inside ``CKEditor.Config`` (advanced config text area) but:
  * ``CKEditor.Config`` doesn't exist by default, so we'd have to provide a default one (either in CKEditor Application or in XS)
  * the advanced config text area doesn't support Velocity code ATM, we'd have to fix this (i.e. https://jira.xwiki.org/browse/XWIKI-12301) if we want to have the default ``CKEditor.Config`` in the CKEditor Application (it needs to detect the support for temporary attachments using Velocity code)
    * alternatively, we could make XS provide a default ``CKEditor.Config`` page, with configuration specific to XS, and this configuration would have in the advanced text area:
      ```
      config['xwiki-upload'].temporaryAttachmentSupported = true;
      ```
  * the Reset action would have to be modified to restore the version provided by the extension (in case ``CKEditor.Config`` is part of an extension, e.g. XS flavor) instead of simply deleting the configuration page

I think the cleanest, on the long term at lest, is to:

* make XS (flavor) provide a default ``CKEditor.Config`` page with configuration specific to XS
  * set ``config['xwiki-upload'].temporaryAttachmentSupported = true;`` in its advanced configuration text area
  * modify the Reset action from ``AdminSection`` to restore the version from the extension rather than deleting the ``CKEditor.Config`` page
* no need to modify ``EditSheet``, ``VelocityMacros`` or ``InlineEditor``
* ``xwiki-upload`` reads this configuration and sets a custom header on each upload request, also listeners to upload response to add the hidden input
* ``FileUploader`` checks the custom HTTP header to know what type of upload to do
* anyone that needs to disable temporary attachment usage can either do this from the instance config, if they control the editor creation, or by listening to editor events and modifying the config just before the editor is fully loaded, as per https://extensions.xwiki.org/xwiki/bin/view/Extension/CKEditor+Integration#HConfiguretheEditor .

Reply:
Thanks for the heads up, I won't go there right now: I'm not even sure that there's a need to disable it globally right now, and as you said we have several mechanism available to disable it where it's needed.

Reply:
Note that an intermediary step in this direction, that removes the need to provide a default ``CKEditor.Config`` and to modify the Reset action, but still doesn't require any changes to ``EditSheet``, ``VelocityMacros`` or ``InlineEditor`` is to modify ``ConfigSheet``, precisely here https://github.com/xwiki-contrib/application-ckeditor/blob/master/application-ckeditor-ui/src/main/resources/CKEditor/ConfigSheet.xml#L59 . We can make this the place to put global config that needs some Velocity code for evaluation. Administrators will still be able to disable temporary attachment easily from the advanced config text area, because that is executed, in JavaScript, on top of the custom config we set in ``ConfigSheet``. Basically:

```
#set ($temporaryAttachmentSupported = "$!services.temporaryAttachments" != '')
## Global default configuration that doesn't depend on the edited content and that needs Velocity code for evaluation.
## Static global configuration (that doesn't need Velocity) should be put in config.js . The default instance level
## configuration is in EditSheet.
#set ($customConfig = {
  'xwiki-upload': {
    'temporaryAttachmentSupported': $temporaryAttachmentSupported
  }
})
```

This means that we'll have this levels of configuration:

* ``config.js``: global **default** (static) configuration, no Velocity evaluation
* ``CKEditor.ConfigSheet``: global **default** (dynamic) configuration, with Velocity evaluation
* ``CKEditor.Config``: global (static) configuration, no Velocity evaluation
* ``CKEditor.EditSheet``: instance **default** (dynamic) configuration
* instance configuration passed when creating the editor
* custom instance configuration done by listening to editor events (from code that doesn't control the editor instance creation)

Reply:
The idea sounds interesting but there's one thing I'm not sure about: it means that users might customize the `CKEditor.ConfigSheet` document, which is not really aimed to be customized, so in case of upgrade it might be problematic no?

Reply:
No, ``CKEditor.ConfigSheet`` is not meant to be customized by users. It provides **default** configuration which can be overwritten from the administration section. I documented this on https://extensions.xwiki.org/xwiki/bin/view/Extension/CKEditor%20Integration/#HConfigurationLevels .

Reply:
Right, so sounds good then.